From a414205e39168ea407c7f5db05e2ca4ca047b6d7 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Fri, 15 Mar 2024 17:26:53 +0100
Subject: [PATCH 0863/1013] evl/stax: add support for spinning wait from the
 in-band stage

Some use cases may require the stax to be acquired from a
non-preemptible in-band context, for which calling schedule() is a
no-go.

Add support for spinning/busy waiting on the stax when attempting to
lock from in-band to enable stage exclusion for such cases. Passing
EVL_STAX_SPIN_INBAND to the new 'flags' parameter of evl_init_stax()
turns this feature on.

The feature is not provided to oob waiters. A spinning wait from
in-band is safe provided oob does not sleep indefinitely while holding
a stax contented by inband. However, preemption of a spinning in-band
waiter by any oob thread waking up from a sleep state on the same CPU
is guaranteed, which keeps the scheme sane.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 drivers/evl/hectic.c |  2 +-
 include/evl/stax.h   |  5 ++++-
 kernel/evl/stax.c    | 37 ++++++++++++++++++++++++++-----------
 3 files changed, 31 insertions(+), 13 deletions(-)

diff --git a/drivers/evl/hectic.c b/drivers/evl/hectic.c
index dbc3b9dc56f2..4cff66cb1e0c 100644
--- a/drivers/evl/hectic.c
+++ b/drivers/evl/hectic.c
@@ -612,7 +612,7 @@ static int hectic_open(struct inode *inode, struct file *filp)
 
 	init_irq_work(&ctx->wake_utask, rtswitch_utask_waker);
 	evl_init_timer(&ctx->wake_up_delay, timed_wake_up);
-	evl_init_stax(&ctx->stax);
+	evl_init_stax(&ctx->stax, 0);
 
 	filp->private_data = ctx;
 	stream_open(inode, filp);
diff --git a/include/evl/stax.h b/include/evl/stax.h
index c8695a87a546..e82f3628dabc 100644
--- a/include/evl/stax.h
+++ b/include/evl/stax.h
@@ -12,14 +12,17 @@
 #include <linux/irq_work.h>
 #include <evl/wait.h>
 
+#define EVL_STAX_INBAND_SPIN  BIT(0)
+
 struct evl_stax {
 	atomic_t gate;
+	int flags;
 	struct evl_wait_queue oob_wait;
 	wait_queue_head_t inband_wait;
 	struct irq_work irq_work;
 };
 
-void evl_init_stax(struct evl_stax *stax);
+void evl_init_stax(struct evl_stax *stax, int flags);
 
 void evl_destroy_stax(struct evl_stax *stax);
 
diff --git a/kernel/evl/stax.c b/kernel/evl/stax.c
index 6cc44f439248..fdba95bd5318 100644
--- a/kernel/evl/stax.c
+++ b/kernel/evl/stax.c
@@ -13,7 +13,10 @@
 #define STAX_INBAND_BIT  BIT(31)
 /*
  * The stax is being claimed by the converse stage. This bit is
- * manipulated exclusively while holding oob_wait.wchan.lock.
+ * manipulated exclusively while holding oob_wait.wchan.lock. When set
+ * from inband, this bit also implicitly means that the waiter is not
+ * spinning but sleeping on the stax (i.e. EVL_STAX_INBAND_SPIN is
+ * clear at init).
  */
 #define STAX_CLAIMED_BIT BIT(30)
 
@@ -22,12 +25,15 @@
 
 static void wakeup_inband_waiters(struct irq_work *work);
 
-void evl_init_stax(struct evl_stax *stax)
+void evl_init_stax(struct evl_stax *stax, int flags)
 {
 	atomic_set(&stax->gate, 0);
+	stax->flags = flags;
 	evl_init_wait(&stax->oob_wait, &evl_mono_clock, EVL_WAIT_FIFO);
-	init_waitqueue_head(&stax->inband_wait);
-	init_irq_work(&stax->irq_work, wakeup_inband_waiters);
+	if (!(flags & EVL_STAX_INBAND_SPIN)) {
+		init_waitqueue_head(&stax->inband_wait);
+		init_irq_work(&stax->irq_work, wakeup_inband_waiters);
+	}
 }
 EXPORT_SYMBOL_GPL(evl_init_stax);
 
@@ -233,17 +239,23 @@ static int lock_from_inband(struct evl_stax *stax, bool wait)
 		 */
 		if (old & STAX_CONCURRENCY_MASK)
 			old |= STAX_INBAND_BIT;
+
 		/* Keep the claim bit in arithmetics. */
 		new = ((old & ~STAX_INBAND_BIT) + 1) | STAX_INBAND_BIT;
 		prev = atomic_cmpxchg(&stax->gate, old, new);
 		if (prev == old)
-			break;
+			break;	/* Acquired. */
+
 		/*
-		 * Retry if the section is still clear for entry from
-		 * inband.
+		 * Cannot acquire for inband yet. Retry if spinning
+		 * wait is requested, or if the section is clear from
+		 * any oob activity.
 		 */
-		if (inband_may_access(prev))
+		if (stax->flags & EVL_STAX_INBAND_SPIN || inband_may_access(prev)) {
+			cpu_relax();
 			continue;
+		}
+
 		if (!wait) {
 			ret = -EAGAIN;
 			break;
@@ -295,7 +307,10 @@ static void unlock_from_oob(struct evl_stax *stax)
 	unsigned long flags;
 	int old, prev, new;
 
-	/* Try the fast path: non-contended (unclaimed by inband). */
+	/*
+	 * Try the fast path: non-contended (unclaimed by inband,
+	 * maybe spinning).
+	 */
 	prev = atomic_read(&stax->gate);
 
 	while (!(prev & STAX_CLAIMED_BIT)) {
@@ -311,8 +326,8 @@ static void unlock_from_oob(struct evl_stax *stax)
 	}
 
 	/*
-	 * stax is claimed by inband, we have to take the slow path
-	 * under lock.
+	 * stax is claimed by inband (which is therefore NOT
+	 * spinning), we have to take the slow path under lock.
 	 */
 	raw_spin_lock_irqsave(&stax->oob_wait.wchan.lock, flags);
 
-- 
2.30.2

